test/extensions/filters/http/common/fuzz/http_filter_fuzzer.h

```diff
@@ -0,0 +1,206 @@
+    end_stream = true;
+  }
+  const auto& headersStatus = sendHeaders(filter, data, end_stream);
+  ENVOY_LOG_MISC(debug, "Finished with FilterHeadersStatus: {}", headersStatus);
+  if ((headersStatus != Http::FilterHeadersStatus::Continue &&
+       headersStatus != Http::FilterHeadersStatus::StopIteration) ||
+      !enabled_) {
+    return;
+  }
+
+  const std::vector<std::string> data_chunks = Fuzz::parseHttpData(data);
+  for (size_t i = 0; i < data_chunks.size(); i++) {
+    if (!data.has_trailers() && i == data_chunks.size() - 1) {
+      end_stream = true;
+    }
+    Buffer::OwnedImpl buffer(data_chunks[i]);
+    const auto& dataStatus = sendData(filter, buffer, end_stream);
+    ENVOY_LOG_MISC(debug, "Finished with FilterDataStatus: {}", dataStatus);
+    if (dataStatus != Http::FilterDataStatus::Continue || !enabled_) {
+      return;
+    }
+  }
+
+  if (data.has_trailers() && enabled_) {
+    sendTrailers(filter, data);
+  }
+}
+
+template <>
+inline Http::FilterHeadersStatus HttpFilterFuzzer::sendHeaders(Http::StreamDecoderFilter* filter,
+                                                               const test::fuzz::HttpData& data,
+                                                               bool end_stream) {
+  request_headers_ = Fuzz::fromHeaders<Http::TestRequestHeaderMapImpl>(data.headers());
+  if (request_headers_.Path() == nullptr) {
+    request_headers_.setPath("/foo");
+  }
+  if (request_headers_.Method() == nullptr) {
+    request_headers_.setMethod("GET");
+  }
+  if (request_headers_.Host() == nullptr) {
+    request_headers_.setHost("foo.com");
+  }
+
+  ENVOY_LOG_MISC(debug, "Decoding headers (end_stream={}):\n{} ", end_stream, request_headers_);
+  Http::FilterHeadersStatus status = filter->decodeHeaders(request_headers_, end_stream);
+  if (end_stream) {
+    filter->decodeComplete();
+  }
+  return status;
+}
+
+template <>
+inline Http::FilterHeadersStatus HttpFilterFuzzer::sendHeaders(Http::StreamEncoderFilter* filter,
+                                                               const test::fuzz::HttpData& data,
+                                                               bool end_stream) {
+  response_headers_ = Fuzz::fromHeaders<Http::TestResponseHeaderMapImpl>(data.headers());
+
+  // Status must be a valid unsigned long. If not set, the utility function below will throw
+  // an exception on the data path of some filters. This should never happen in production, so catch
+  // the exception and set to a default value.
+  try {
+    (void)Http::Utility::getResponseStatus(response_headers_);
+  } catch (const Http::CodecClientException& e) {
+    response_headers_.setStatus(200);
+  }
+
+  ENVOY_LOG_MISC(debug, "Encoding headers (end_stream={}):\n{} ", end_stream, response_headers_);
+  Http::FilterHeadersStatus status = filter->encodeHeaders(response_headers_, end_stream);
+  if (end_stream) {
+    filter->encodeComplete();
+  }
+  return status;
+}
+
+template <>
```

Comment:
Why are these all templated?

Reply:
The logic of handling filter return status and end_stream values in `runData` is the common for encoders/decoder, so we meant to share that across both encoder/decoder despite different methods.

There were some alternatives considered here https://github.com/envoyproxy/envoy/pull/11209, one of which was adding a lambda in for the correct header/data/trailer funcs (encodeHeaders, decodeHeaders), and having the `runData` use those lambda HeaderFunc, DataFunc, TrailerFunc depending on if it's encoder or decoder. 

If I were to de-template this, it's not too much additional work. The lambdas (Header/Data/Trailer)Funcs are the template send(Header/Data/Trailer) specializations.  I guess it also depends on what API is exposed to the fuzzer:
`runData(FilterType, data)` vs `runEncode(encoder, data)` and `runDecode(decoder, data)` 
where `runEncode(encoder, data)` calls `runData` with the appropriate lambas. I was happy with both.

Reply:
quick ping?

Reply:
FYI if we don't template these functions, then you have to duplicate all the code in `runData`, right?

https://github.com/envoyproxy/envoy/blob/d60073c31355857b5f4bd1b793c478e0fcd82795/test/extensions/filters/http/common/fuzz/http_filter_fuzzer.h#L84-L114

Reply:
I don't think so. runData's signature could be 
`runData(FilterType* filter, const test::fuzz::HttpData& data, HeaderFunc header_func, DataFunc data_func, TrailerFunc trailer_func)` 
and `runEncode(encoder, data)` would call runData with the right funcs.